Restructure title/message and add docker compose context and service names (if available)

DL6ER opened the pull request:
As a base for further evolution on ...

When running `docker compose restart` for `unbound`:

![Screenshot from 2023-09-23 17-32-44](https://github.com/yubiuser/docker-event-monitor/assets/16748619/f148d7df-b5ca-42b7-8143-cf1bb3ee8636)

Exemplary mail:

![Screenshot from 2023-09-23 17-32-54](https://github.com/yubiuser/docker-event-monitor/assets/16748619/7891bf63-7e5a-4eda-9de9-23f75c6c31df)

and on mobile with K-9 mail:

![signal-2023-09-23-174346](https://github.com/yubiuser/docker-event-monitor/assets/16748619/9e3da4b5-5de2-4adb-ab78-84d7ae9f856a)

yubiuser reviewed (changes requested):
I reviewed only the startup changes (commit 2) so far. Suggesting these changes to a) use more precise variable name b) re-order a bit c) use a string builder to save some bytes (go stings are immutable and each `+=` creates a new one)

```go
	log.Info("Starting docker event monitor")

	var startup_message strings.Builder

	if glb_arguments.Pushover {
		startup_message.WriteString("Notify via Pushover, using API Token " + glb_arguments.PushoverAPIToken + " and user key " + glb_arguments.PushoverUserKey)
	} else {
		startup_message.WriteString("Pushover notification disabled")
	}

	if glb_arguments.Gotify {
		startup_message.WriteString("\nNotify via Gotify, using URL " + glb_arguments.GotifyURL + " and APP Token " + glb_arguments.GotifyToken)
	} else {
		startup_message.WriteString("\nGotify notification disabled")
	}
	if glb_arguments.Mail {
		startup_message.WriteString("\nNotify via E-Mail from " + glb_arguments.MailFrom + " to " + glb_arguments.MailTo + " using host " + glb_arguments.MailHost + " and port " + strconv.Itoa(glb_arguments.MailPort))
	} else {
		startup_message.WriteString("\nE-Mail notification disabled")
	}

	if glb_arguments.Delay > 0 {
		startup_message.WriteString("\nUsing delay of " + glb_arguments.Delay.String())
	} else {
		startup_message.WriteString("\nDelay disabled")
	}

	log.Info(startup_message.String())
	sendNotifications(startup_message.String(), "Starting docker event monitor")

	filterArgs := filters.NewArgs()
	for key, values := range glb_arguments.Filter {
		for _, value := range values {
			filterArgs.Add(key, value)
		}
	}
```

yubiuser commented on src/docker-event-monitor.go:
```diff
@@ -286,33 +290,55 @@ func processEvent(event *events.Message) {
+	}
+	title += ": " + event.Action
+
+	// Start message with human readable time
```
There is no time within the message

DL6ER commented:
Thanks. I addresses your changes, bear with me, this is the first Go code I'm ever looking at in my life :slightly_smiling_face: 

I also added the `Date` field to the email header. Doing so in a separate PR would have caused merge conflict down the line due to the switch from `+=` to `strings.Builder` here, too., so I did it in-place. There was a funny incident this morning: When I switched on my laptop, I noticed that all notifications overnight appeared at 07:22 whereas my 24/7 running smartphone had them at the right time. The reason was the lack of a `Date` field which allows email clients to substitute the time of receipt (so when the laptop was switched on) as time in lack of anything better.

We could possibly further add
```
msg.WriteString("Content-Type: text/plain; charset=UTF-8\r\n")
```
but I don't think this is really necessary.

yubiuser commented:
> I also added the Date field to the email header. 

Good idea. I wonder if it's worth using the event's timestamp instead of the time the message is build. There might be a few seconds delay depending on the number of events. 
We could also add the timestamp to `processEvent()` to send it within the message as well?
Factor out building of the startup message? It got quite long within `main()`?

yubiuser commented:
Adding the content type as well is a good idea.

yubiuser commented:
I pushed 3 commits with refinements to your branch. This was easier than to explain the changes I had in mind. If you're fine with this, we can merge this

DL6ER commented:
I noticed the first mail having a different title, but this may have been like that before. Question: Do we want it like that?
![Screenshot from 2024-01-24 17-02-45](https://github.com/yubiuser/docker-event-monitor/assets/16748619/d268b88a-0d7c-408e-be5c-b1c64b263216)

First message content:
```
image: gitea/gitea:latest
Name: d521127acb9b_gitea                           <== There we have the strange name
Time: Wed, 24 Jan 2024 17:01:49 +0100
Docker compose context: /home/me/docker/gitea
Docker compose service: server
```
Second message:
```
image: gitea/gitea:latest
Name: gitea                                        <== Here the name is as expected
Time: Wed, 24 Jan 2024 17:01:49 +0100
Docker compose context: /home/me/docker/gitea
Docker compose service: server
```

DL6ER commented:
Turns out I already had `debug` logging on during the mails above:
```
docker-event-monitor  | level=info msg="Starting docker event monitor"
docker-event-monitor  | level=info msg="Docker event monitor started at Wed, 24 Jan 2024 17:01:23 +0100\nPushover notification disabled\nGotify notification disabled\nNotify via E-Mail from docker-event-monitor@xxx.de to d@xxx.de using host mail.xxx.de and port 587\nUsing delay of 500ms\nLog level: debug"
docker-event-monitor  | level=debug msg="filterArgs = {map[event:map[create:true die:true kill:true start:true stop:true] type:map[container:true]]}"
docker-event-monitor  | level=debug msg="&events.Message{Status:\"create\", ID:\"95d98d5f3ddc8303c2264d45f3277e4e86c798d87e654858cba091f7e37bae97\", From:\"gitea/gitea:latest\", Type:\"container\", Action:\"create\", Actor:events.Actor{ID:\"95d98d5f3ddc8303c2264d45f3277e4e86c798d87e654858cba091f7e37bae97\", Attributes:map[string]string{\"com.docker.compose.config-hash\":\"7d6aeddcdfd43cc48aa8cae09e522f7f849c95657e23a4c63cd3b129c8d050bf\", \"com.docker.compose.container-number\":\"1\", \"com.docker.compose.depends_on\":\"db:service_started:false\", \"com.docker.compose.image\":\"sha256:ea5f0390c8805804ed5dcd325b970efa681446d327f022ee84304de592cfc6dd\", \"com.docker.compose.oneoff\":\"False\", \"com.docker.compose.project\":\"gitea\", \"com.docker.compose.project.config_files\":\"/home/me/docker/gitea/docker-compose.yml\", \"com.docker.compose.project.working_dir\":\"/home/me/docker/gitea\", \"com.docker.compose.replace\":\"d521127acb9bc50ace37957f65aa7eafdc88237027d5eafe8a8825b401422d9b\", \"com.docker.compose.service\":\"server\", \"com.docker.compose.version\":\"2.21.0\", \"image\":\"gitea/gitea:latest\", \"maintainer\":\"maintainers@gitea.io\", \"name\":\"d521127acb9b_gitea\", \"org.opencontainers.image.created\":\"2024-01-17T02:49:10.805Z\", \"org.opencontainers.image.description\":\"Git with a cup of tea! Painless self-hosted all-in-one software development service, including Git hosting, code review, team collaboration, package registry and CI/CD\", \"org.opencontainers.image.licenses\":\"MIT\", \"org.opencontainers.image.revision\":\"f4923854f660c75f75592b9506c0471475ecc7cf\", \"org.opencontainers.image.source\":\"https://github.com/go-gitea/gitea\", \"org.opencontainers.image.title\":\"gitea\", \"org.opencontainers.image.url\":\"https://github.com/go-gitea/gitea\", \"org.opencontainers.image.version\":\"1\"}}, Scope:\"local\", Time:1706112109, TimeNano:1706112109212142120}"
docker-event-monitor  | level=info msg="Container d521127acb9b_gitea: createimage: gitea/gitea:latest\nName: d521127acb9b_gitea\nTime: Wed, 24 Jan 2024 17:01:49 +0100\nDocker compose context: /home/me/docker/gitea\nDocker compose service: server"
docker-event-monitor  | level=debug msg="&events.Message{Status:\"kill\", ID:\"d521127acb9bc50ace37957f65aa7eafdc88237027d5eafe8a8825b401422d9b\", From:\"gitea/gitea:latest\", Type:\"container\", Action:\"kill\", Actor:events.Actor{ID:\"d521127acb9bc50ace37957f65aa7eafdc88237027d5eafe8a8825b401422d9b\", Attributes:map[string]string{\"com.docker.compose.config-hash\":\"7d6aeddcdfd43cc48aa8cae09e522f7f849c95657e23a4c63cd3b129c8d050bf\", \"com.docker.compose.container-number\":\"1\", \"com.docker.compose.depends_on\":\"db:service_started:false\", \"com.docker.compose.image\":\"sha256:b0bd40c24dcdddbcabfa4b5a5ae87c9c80cc6c1ea556d187549cb112f1666757\", \"com.docker.compose.oneoff\":\"False\", \"com.docker.compose.project\":\"gitea\", \"com.docker.compose.project.config_files\":\"/home/me/docker/gitea/docker-compose.yml\", \"com.docker.compose.project.working_dir\":\"/home/me/docker/gitea\", \"com.docker.compose.replace\":\"17f6a56322a12267cc1d3e2763924c32dbdb0b13074e8fd5524ed9d49a0d48b2\", \"com.docker.compose.service\":\"server\", \"com.docker.compose.version\":\"2.21.0\", \"image\":\"gitea/gitea:latest\", \"maintainer\":\"maintainers@gitea.io\", \"name\":\"gitea\", \"org.opencontainers.image.created\":\"2023-12-21T08:32:46.683Z\", \"org.opencontainers.image.description\":\"Git with a cup of tea! Painless self-hosted all-in-one software development service, including Git hosting, code review, team collaboration, package registry and CI/CD\", \"org.opencontainers.image.licenses\":\"MIT\", \"org.opencontainers.image.revision\":\"2360c7ec6cdddc337a9b4964a64d2e1640c69dad\", \"org.opencontainers.image.source\":\"https://github.com/go-gitea/gitea\", \"org.opencontainers.image.title\":\"gitea\", \"org.opencontainers.image.url\":\"https://github.com/go-gitea/gitea\", \"org.opencontainers.image.version\":\"1\", \"signal\":\"15\"}}, Scope:\"local\", Time:1706112109, TimeNano:1706112109222071479}"
docker-event-monitor  | level=info msg="Container gitea: killimage: gitea/gitea:latest\nName: gitea\nTime: Wed, 24 Jan 2024 17:01:49 +0100\nDocker compose context: /home/me/docker/gitea\nDocker compose service: server"
docker-event-monitor  | level=debug msg="&events.Message{Status:\"stop\", ID:\"d521127acb9bc50ace37957f65aa7eafdc88237027d5eafe8a8825b401422d9b\", From:\"gitea/gitea:latest\", Type:\"container\", Action:\"stop\", Actor:events.Actor{ID:\"d521127acb9bc50ace37957f65aa7eafdc88237027d5eafe8a8825b401422d9b\", Attributes:map[string]string{\"com.docker.compose.config-hash\":\"7d6aeddcdfd43cc48aa8cae09e522f7f849c95657e23a4c63cd3b129c8d050bf\", \"com.docker.compose.container-number\":\"1\", \"com.docker.compose.depends_on\":\"db:service_started:false\", \"com.docker.compose.image\":\"sha256:b0bd40c24dcdddbcabfa4b5a5ae87c9c80cc6c1ea556d187549cb112f1666757\", \"com.docker.compose.oneoff\":\"False\", \"com.docker.compose.project\":\"gitea\", \"com.docker.compose.project.config_files\":\"/home/me/docker/gitea/docker-compose.yml\", \"com.docker.compose.project.working_dir\":\"/home/me/docker/gitea\", \"com.docker.compose.replace\":\"17f6a56322a12267cc1d3e2763924c32dbdb0b13074e8fd5524ed9d49a0d48b2\", \"com.docker.compose.service\":\"server\", \"com.docker.compose.version\":\"2.21.0\", \"image\":\"gitea/gitea:latest\", \"maintainer\":\"maintainers@gitea.io\", \"name\":\"gitea\", \"org.opencontainers.image.created\":\"2023-12-21T08:32:46.683Z\", \"org.opencontainers.image.description\":\"Git with a cup of tea! Painless self-hosted all-in-one software development service, including Git hosting, code review, team collaboration, package registry and CI/CD\", \"org.opencontainers.image.licenses\":\"MIT\", \"org.opencontainers.image.revision\":\"2360c7ec6cdddc337a9b4964a64d2e1640c69dad\", \"org.opencontainers.image.source\":\"https://github.com/go-gitea/gitea\", \"org.opencontainers.image.title\":\"gitea\", \"org.opencontainers.image.url\":\"https://github.com/go-gitea/gitea\", \"org.opencontainers.image.version\":\"1\"}}, Scope:\"local\", Time:1706112110, TimeNano:1706112110598488897}"
docker-event-monitor  | level=info msg="Container gitea: stopimage: gitea/gitea:latest\nName: gitea\nTime: Wed, 24 Jan 2024 17:01:50 +0100\nDocker compose context: /home/me/docker/gitea\nDocker compose service: server"
docker-event-monitor  | level=debug msg="&events.Message{Status:\"die\", ID:\"d521127acb9bc50ace37957f65aa7eafdc88237027d5eafe8a8825b401422d9b\", From:\"gitea/gitea:latest\", Type:\"container\", Action:\"die\", Actor:events.Actor{ID:\"d521127acb9bc50ace37957f65aa7eafdc88237027d5eafe8a8825b401422d9b\", Attributes:map[string]string{\"com.docker.compose.config-hash\":\"7d6aeddcdfd43cc48aa8cae09e522f7f849c95657e23a4c63cd3b129c8d050bf\", \"com.docker.compose.container-number\":\"1\", \"com.docker.compose.depends_on\":\"db:service_started:false\", \"com.docker.compose.image\":\"sha256:b0bd40c24dcdddbcabfa4b5a5ae87c9c80cc6c1ea556d187549cb112f1666757\", \"com.docker.compose.oneoff\":\"False\", \"com.docker.compose.project\":\"gitea\", \"com.docker.compose.project.config_files\":\"/home/me/docker/gitea/docker-compose.yml\", \"com.docker.compose.project.working_dir\":\"/home/me/docker/gitea\", \"com.docker.compose.replace\":\"17f6a56322a12267cc1d3e2763924c32dbdb0b13074e8fd5524ed9d49a0d48b2\", \"com.docker.compose.service\":\"server\", \"com.docker.compose.version\":\"2.21.0\", \"execDuration\":\"32211\", \"exitCode\":\"0\", \"image\":\"gitea/gitea:latest\", \"maintainer\":\"maintainers@gitea.io\", \"name\":\"gitea\", \"org.opencontainers.image.created\":\"2023-12-21T08:32:46.683Z\", \"org.opencontainers.image.description\":\"Git with a cup of tea! Painless self-hosted all-in-one software development service, including Git hosting, code review, team collaboration, package registry and CI/CD\", \"org.opencontainers.image.licenses\":\"MIT\", \"org.opencontainers.image.revision\":\"2360c7ec6cdddc337a9b4964a64d2e1640c69dad\", \"org.opencontainers.image.source\":\"https://github.com/go-gitea/gitea\", \"org.opencontainers.image.title\":\"gitea\", \"org.opencontainers.image.url\":\"https://github.com/go-gitea/gitea\", \"org.opencontainers.image.version\":\"1\"}}, Scope:\"local\", Time:1706112110, TimeNano:1706112110600053227}"
docker-event-monitor  | level=info msg="Container gitea: dieimage: gitea/gitea:latest\nName: gitea\nTime: Wed, 24 Jan 2024 17:01:50 +0100\nDocker compose context: /home/me/docker/gitea\nDocker compose service: server"
docker-event-monitor  | level=debug msg="&events.Message{Status:\"start\", ID:\"95d98d5f3ddc8303c2264d45f3277e4e86c798d87e654858cba091f7e37bae97\", From:\"gitea/gitea:latest\", Type:\"container\", Action:\"start\", Actor:events.Actor{ID:\"95d98d5f3ddc8303c2264d45f3277e4e86c798d87e654858cba091f7e37bae97\", Attributes:map[string]string{\"com.docker.compose.config-hash\":\"7d6aeddcdfd43cc48aa8cae09e522f7f849c95657e23a4c63cd3b129c8d050bf\", \"com.docker.compose.container-number\":\"1\", \"com.docker.compose.depends_on\":\"db:service_started:false\", \"com.docker.compose.image\":\"sha256:ea5f0390c8805804ed5dcd325b970efa681446d327f022ee84304de592cfc6dd\", \"com.docker.compose.oneoff\":\"False\", \"com.docker.compose.project\":\"gitea\", \"com.docker.compose.project.config_files\":\"/home/me/docker/gitea/docker-compose.yml\", \"com.docker.compose.project.working_dir\":\"/home/me/docker/gitea\", \"com.docker.compose.replace\":\"d521127acb9bc50ace37957f65aa7eafdc88237027d5eafe8a8825b401422d9b\", \"com.docker.compose.service\":\"server\", \"com.docker.compose.version\":\"2.21.0\", \"image\":\"gitea/gitea:latest\", \"maintainer\":\"maintainers@gitea.io\", \"name\":\"gitea\", \"org.opencontainers.image.created\":\"2024-01-17T02:49:10.805Z\", \"org.opencontainers.image.description\":\"Git with a cup of tea! Painless self-hosted all-in-one software development service, including Git hosting, code review, team collaboration, package registry and CI/CD\", \"org.opencontainers.image.licenses\":\"MIT\", \"org.opencontainers.image.revision\":\"f4923854f660c75f75592b9506c0471475ecc7cf\", \"org.opencontainers.image.source\":\"https://github.com/go-gitea/gitea\", \"org.opencontainers.image.title\":\"gitea\", \"org.opencontainers.image.url\":\"https://github.com/go-gitea/gitea\", \"org.opencontainers.image.version\":\"1\"}}, Scope:\"local\", Time:1706112110, TimeNano:1706112110857634281}"
docker-event-monitor  | level=info msg="Container gitea: startimage: gitea/gitea:latest\nName: gitea\nTime: Wed, 24 Jan 2024 17:01:50 +0100\nDocker compose context: /home/me/docker/gitea\nDocker compose service: server"
```

In my example above, `gitea` was updated upstream so `compose` recreated the container on `up -d`. However, I can also reliably recreate this at any point using `docker compose up -d --force-recreate`

yubiuser commented:
I think it's fine as it is. The interesting line from your log is number 4, containing the relevant `&events.Message`. When you look carefully, you see that `docker` reports `events.Actor.Attribute["name"]` as `d521127acb9b_gitea`. We pick this name rightfully. You need to talk to docker about this ;-)
